### PR TITLE
MGMT-16629: Use custom query params and headers for OS image downloads

### PR DESCRIPTION
**Pull request by paul-maidment**
## Description
This PR provides the ability to add custom query parameters and headers to every http request made while downloading OS images.

Two environment variables are to be provided `OS_IMAGES_HTTP_REQUEST_QUERY_PARAMS` and `OS_IMAGES_HTTP_REQUEST_HEADERS` These variables are JSON representations of `map[string]string` where the key is the parameter (or header) name and the value is the parameter (or header) value.

Once provided, these values will be used in the HTTP request headers and the HTTP query parameters of any request made during the download of OS images.

## Testing
This code includes tests to ensure that the functionality works as intended. A manual test will also be performed in podman to ensure the behavior is as expected

## Assignees
/cc @carbonin 
/cc @omertuc 

## Links
This is part of the Epic https://issues.redhat.com/browse/MGMT-15408

## Checklist

- [x] Title and description added to both, commit and PR
- [x] Relevant issues have been associated
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit tests (note that code changes require unit tests)

### CODE REVIEW

File: main.go
Comment (omertuc) [inline main.go:53]:
```suggestion
	// OSImageDownloadHeaders contains a JSON encoded representation of any
	// HTTP headers to be sent with every request to download an os image.

```

File: main.go
Comment (omertuc) [inline main.go:55]:
```suggestion
	// OSImageDownloadQueryParams contains a JSON encoded representation of any
	// query parameters to be sent with every request to download an OS image.
```

File: main.go
Comment (omertuc) [inline main.go:63]:
```suggestion
		if err := json.Unmarshal([]byte(jsonMap), &result); err != nil {
```

File: pkg/imagestore/imagestore.go
Comment (omertuc) [inline pkg/imagestore/imagestore.go:90]:
```suggestion
func NewImageStore(ed isoeditor.Editor, dataDir, imageServiceBaseURL string, insecureSkipVerify bool, versions []map[string]string,
    osImageDownloadTrustedCAFile string, osImageDownloadHeadersMap map[string]string, osImageDownloadQueryParamsMap map[string]string) (ImageStore, error) {
```

File: pkg/imagestore/imagestore.go
Comment (omertuc) [inline pkg/imagestore/imagestore.go:154]: I find `getHttpResponse` to be an odd name for something that makes an HTTP request, I think if we call it `doHttpRequest` or `performHTTPRequest` or something, the response return value is already implied

File: pkg/imagestore/imagestore.go
Comment (omertuc) [inline pkg/imagestore/imagestore.go:159]: Comment unnecessary
```suggestion
```

File: pkg/imagestore/imagestore.go
Comment (omertuc) [inline pkg/imagestore/imagestore.go:163]: Comment unnecessary

```suggestion
```

File: pkg/imagestore/imagestore.go
Comment (omertuc) [inline pkg/imagestore/imagestore.go:171]: Comment unnecessary

```suggestion
```

File: pkg/imagestore/imagestore.go
Comment (omertuc) [inline pkg/imagestore/imagestore.go:157]: Wrap the error

File: pkg/imagestore/imagestore.go
Comment (omertuc) [inline pkg/imagestore/imagestore.go:174]: Wrap the error

**Review by omertuc — CHANGES_REQUESTED** [review]
Looks great apart for some cosmetic nits

File: main.go
Comment (omertuc) [inline main.go:54]:
```suggestion
	// HTTP headers to be sent with every request to download an OS image.
```

File: pkg/imagestore/imagestore.go
Comment (omertuc) [inline pkg/imagestore/imagestore.go:181]:
```suggestion
		return fmt.Errorf(err, "http request to %s failed: %w", url, err)
```

**Review by omertuc — CHANGES_REQUESTED** [review]

**Comment by carbonin** [comment]
/hold

I have a few comments

File: main.go
Comment (carbonin) [inline main.go:55]: Maybe the env var name should match the variable name here?

File: main.go
Comment (carbonin) [inline main.go:61]: I thought envconfig handled maps? Did you look into that?

**Review by carbonin — COMMENTED** [review]
Some rather minor comments.

If envconfig supports json maps I'd like to use that over deserializing manually

File: main.go
Comment (carbonin) [reply main.go:61]: Ah, answered my own question. It doesn't quite use json  https://github.com/kelseyhightower/envconfig/blob/v1.4.0/envconfig.go#L319-L342

Probably best to stick with json then.
Also I suppose we're doing this already for OS Images :shrug:

**Comment by carbonin** [comment]
Okay given that the only comment is variable naming I'll leave it up to you @paul-maidment if you want to change that.
If not just remove the hold.

**Comment by paul-maidment** [comment]
> Okay given that the only comment is variable naming I'll leave it up to you @paul-maidment if you want to change that. If not just remove the hold.

I renamed the variable and I also changed the way we bring the map in, so that we use envconfig directly for that.

File: main.go
Comment (paul-maidment) [reply main.go]: Yeah, just changed it now to get rid of this

File: main.go
Comment (paul-maidment) [reply main.go]: Done

**Comment by carbonin** [comment]
I think the env var processing needs to go back to the way it was.
envconfig doesn't actually parse json for maps it seems

**Review by carbonin — APPROVED** [review]

**Pull request merged**
